src/librustc_mir/interpret/eval_context.rs

```diff
@@ -558,56 +560,142 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
+                    // return place is always a local and then this cannot happen.
+                    self.validate_operand(
+                        self.place_to_op(return_place)?,
+                        vec![],
+                        None,
+                    )?;
+                }
+            } else {
+                // Uh, that shouldn't happen... the function did not intend to return
+                throw_ub!(Unreachable);
+            }
+
+            // Jump to new block -- *after* validation so that the spans make more sense.
+            match frame.return_to_block {
+                StackPopCleanup::Goto { ret, ..  } => {
+                    self.goto_block(ret)?;
+                }
+                StackPopCleanup::None { .. } => {}
             }
-        } else {
-            // Uh, that shouldn't happen... the function did not intend to return
-            throw_ub!(Unreachable)
         }
-        // Jump to new block -- *after* validation so that the spans make more sense.
-        match frame.return_to_block {
-            StackPopCleanup::Goto(block) => {
-                self.goto_block(block)?;
+
+
+        Ok((frame.return_to_block, stack_pop_info))
+    }
+
+    pub(super) fn pop_stack_frame(&mut self, unwinding: bool) -> InterpResult<'tcx> {
+        let (mut cleanup, mut stack_pop_info) = self.pop_stack_frame_internal(unwinding)?;
+
+        // There are two cases where we want to unwind the stack:
+        // * The caller explicitly told us (i.e. we hit a Resume terminator)
+        // * The machine indicated that we've just started unwinding (i.e.
+        // a panic has just occured)
+        if unwinding || stack_pop_info == StackPopInfo::StartUnwinding {
+            trace!("unwinding: starting stack unwind...");
+            // Overwrite our current stack_pop_info, so that the check
+            // below doesn't fail.
+            stack_pop_info = StackPopInfo::Normal;
+            // There are three posible ways that we can exit the loop:
+            // 1) We find an unwind block - we jump to it to allow cleanup
+            // to occur for that frame
+            // 2) pop_stack_frame_internal reports that we're no longer unwinding
+            //    - this means that the panic has been caught, and that execution
+            //    should continue as normal
+            // 3) We pop all of our frames off the stack - this should never happen.
+            while !self.stack.is_empty() {
+                match stack_pop_info {
+                    // We tried to start unwinding while we were already
+                    // unwinding. Note that this **is not** the same thing
+                    // as a double panic, which will be intercepted by
+                    // libcore/libstd before we actually attempt to unwind.
+                    StackPopInfo::StartUnwinding => {
+                        throw_ub_format!("Attempted to start unwinding while already unwinding!");
```

Comment:
Is this actually reachable? I think this would only be possible with broken MIR, right?
If so, just `bug!` out here

Reply:
That won't print a backtrace for the interpreted program, right? If this is ever hit, it will be useful to know where to start looking.

Reply:
Hmm... we could have a special error type for MIR bugs I guess, but you can also use `span_bug` and point to the current statement

Reply:
If this is supposedly impossible, "invalid program" might be a better error class.

But UB also seems fine for me.

Reply:
> But UB also seems fine for me.

Okay here I was thinking this could be reached by calling the panic runtime in a bad way during unwinding... can it?

Reply:
No - [libstd prevents this](https://github.com/rust-lang/rust/blob/89e645ace853d86e8c0002247482eac9073c90fd/src/libstd/panicking.rs#L443-L454). This line should only be hit if there's a bug in libstd.

Reply:
Well that's enough though I think to not make it ICE -- we don't have a strict policy for these cases but I think making it UB is reasonable. libstd might have bugs or people might run Miri with custom libstd.